protocols/identify/src/behaviour.rs

```diff
@@ -670,6 +691,39 @@ impl PeerCache {
     }
 }
 
+#[derive(Debug, Clone)]
+#[allow(clippy::large_enum_variant)]
+pub(crate) enum KeyType {
+    // With public key only the behaviour will not
+    // be able to produce a `SignedEnvelope`.
+    PublicKey(PublicKey),
+    Keypair {
+        keypair: Keypair,
+        public_key: PublicKey,
+    },
+}
```

Comment:
Would it make sense to use `Box<Keypair>` here to avoid the large-enum-variant issue?

Reply:
I don't have an opinion on this. Might not be necessary since only the `Behaviour` holds one instance of this enum.

Reply:
Since we are duplicating this in every `ConnectionHandler`, I think we can reference counting it. What do you think @elenaf9?

Reply:
Yes sounds good :+1:

Reply:
In that case do we still box the data inside? If so, would it be better if we use a tuple or create a struct for semantic purpose?

Reply:
> In that case do we still box the data inside?

No need to `Box` it, `Arc` already allocates the data on the heap.

Reply:
> No need to `Box` it, `Arc` already allocates the data on the heap.

I know but the `Box` was supposed to get rid of `large_enum_variant` right? Does it make sense to have two separate `Arc`?

Reply:
> I know but the `Box` was supposed to get rid of `large_enum_variant` right? Does it make sense to have two separate `Arc`?

I am not sure I know what you mean.
My idea was to just make the `keypair: Keypair` propterty a `Box<Keypair>`/ `Arc<Keypair>` and leave `public_key: PublicKey` as it is. That way the keypair would be on the heap and the enum only stores the pointer, thus avoiding the `large_enum_variant`.

Alternatively we could also not have this enum, stick with the old `local_public_key: PublicKey` and add have an additional `records_signing_key: Option<Keypair>`.

Reply:
I think it might be best to go for the alt Elena mentioned if we are to avoid using Box/Arc. Additionally if we are to use it, we could use a struct that as an inner enum that is wrapped with `Arc` but at that point it would probably be better to just have a struct with two fields with one being a public key and another an optional keypair (though at that point we are getting back to what was mentioned above).

Reply:
Let me rephrase. First, we would like to get rid of `large_enum_variant` lint using heap allocation. Then I tried to box all fields(misread your message) but I didn't think allocate twice is very optimal so I asked should we do `Box<BothKey>` or `Box<(Keypair, PublicKey)>`. Plus we are duplicating `KeyType` the enum in every `ConnectionHandler`, so I asked "should we reference counting it(`KeyType`)" so that there would only be one copy of `KeyType` and the lint won't really matter. I guess what you mean is reference counting one field but still cloning `KeyType`. I think it would be better if we just reference count the entire enum.

Reply:
> I think it would be better if we just reference count the entire enum

Fine by me as well. I think the increase enum size is okay if it's only allocated once.